### PR TITLE
Use multi-stage build and replace CMD with ENTRYPOINT

### PR DESCRIPTION
**Pull request by dliappis**
This PR moves away from running the `es-docker` wrapper script via `CMD` to a new `ENTRYPOINT` script.

The `ENTRYPOINT` runs as root,  on vanilla docker and most orchestration frameworks, and switches to the pre-provisioned `elasticsearch` user with uid=gid=1000 before exec'ing `elasticsearch`.

It's still possible to run another `CMD`, through the `ENTRYPOINT`, such
as `/bin/bash` to inspect things or directly `elasticsearch -Ex.y=z`[4].

As there is at least one docker orchestration framework, Openshift Open[1], that overrides the Dockerfile `USER` to something with random uid and gid=0, the entrypoint script conditionally skips the privilege de-escalation in this case. To achieve this, the `elasticsearch.tar.gz` needs to be extracted so that all files+dirs are gid=0 and have the same group permissions as the owner. This is according to RedHat's best practices described in [1].

We avoid generating another large layer due to `chown -R` and `chgrp -R`, by utilizing multi-stage builds and the new `COPY --chown=` feature[2], available starting with `docker-ce >=17.09`. This is only a build requirement, that doesn't affect the portability of the image. As a side effect we are also able to significantly clean up the Dockerfile jinja2 template.

We also introduce the optional `TAKE_FILE_OWNERSHIP` flag which will. in most orchestration environments, forcefully change the uid and gid of the data and logdir. This can help users bind-mounting host directories that don't provide write access either to the default user 1000 or to gid 0. This flag is not expected to receive much use in orchestration frameworks as k8s has the securityContext option[3], elsewhere volumes are typically gid=0 and g=rwx and docker+docker swarm have named volumes and storage drivers.

The PR also includes three new tests:

1) ensure all extracted files under `/usr/share/elasticsearch` have the right gid.
2) tests that elasticsearch works ok when using a bind-mounted host directory for data.
3) tests that elasticsearch works ok also when a bind-mounted host dir has a random uid (but gid=0).

There are a number of more tests that we can add here (e.g. for the `MUTATE_MOUNTS` flag, testing that elasticsearch can start with `--user` forced to a different value), but I wanted to keep the size of this PR under control for now and add tests, if needed, in an additional PR.

\[1\]: https://docs.openshift.org/latest/creating_images/guidelines.html
\[2\]: https://github.com/moby/moby/pull/34263
\[3\]: kubernetes/kubernetes#2630 (comment)
\[4\]: https://www.elastic.co/guide/en/elasticsearch/reference/current/docker.html#_d_override_the_image_8217_s_default_ulink_url_https_docs_docker_com_engine_reference_run_cmd_default_command_or_options_cmd_ulink
f674bf9

Closes: #86, #90, #114,

### CODE REVIEW

**Comment by dliappis** [comment]
@drewr FYI as discussed

cc @xeraa 

@agc93 This PR should resolve the issue #114 you reported using Openshift Open.

**Comment by dliappis** [comment]
Also FYI @tylerjl

**Comment by agc93** [comment]
@dliappis That's awesome, thanks!

File: build/elasticsearch/bin/docker-entrypoint.sh [hunk 1]
```diff
@@ -1,6 +1,40 @@
...
 
-# Run Elasticsearch and allow setting default settings via env vars
+# Files created by Elasticsearch should always be group writable too
+umask 0002
+
+# Create random user entry, if different from uid 1000, to support Openshift
+# https://docs.openshift.org/latest/creating_images/guidelines.html
+if ! whoami &> /dev/null; then
+    if [ -w /etc/passwd ]; then
+        echo "${USER_NAME:-default}:x:$(id -u):0:${USER_NAME:-default} user:/usr/share/elasticsearch:/sbin/nologin" >> /etc/passwd
```
Comment (jarpy) [inline build/elasticsearch/bin/docker-entrypoint.sh]: I don't think this user should be called `default`. The default user is `elasticsearch`. This user is the opposite of default. It's an unusual, unexpected and custom user. Let's call it `custom_user` or something.

From a different angle, did you consider changing the UID of the existing `elasticsearch` user instead? Is that a bad idea?

File: build/elasticsearch/bin/docker-entrypoint.sh [hunk 2]
```diff
@@ -1,6 +1,40 @@
...
+        exec "${@}"
+    fi
+}
+
+# Allow user specify custom CMD, maybe bin/elasticsearch itself
+# for example to directly specify `-E` style parameters for elasticsearch on k8s
+# or simply to run /bin/bash to check the image
+if [[ "$1" != "eswrapper" ]]; then
+    if [[ "$(id -u)" == "0" ]] && [[ "$1" == *elasticsearch* ]]; then
+        exec chroot --userspec=1000 / "$@"
```
Comment (jarpy) [inline build/elasticsearch/bin/docker-entrypoint.sh]: Can we get a comment about this very unusual use of chroot? (It's very clever but very weird).

File: templates/Dockerfile.j2 [hunk 3]
```diff
@@ -62,60 +71,74 @@ RUN for PLUGIN in {{ ingest_plugins }}; do \
...
 
-CMD ["/bin/bash", "bin/es-docker"]
+RUN groupadd -g 1000 elasticsearch && \
+    adduser -u 1000 -g 1000 -G 0 -d /usr/share/elasticsearch elasticsearch && \
+    chmod 0775 /usr/share/elasticsearch && \
+    chgrp 0 /usr/share/elasticsearch
+
+WORKDIR /usr/share/elasticsearch
+
+COPY --from=prep_es_files --chown=1000:0 /usr/share/elasticsearch /usr/share/elasticsearch
```
Comment (jarpy) [inline templates/Dockerfile.j2]: Yay!!!!!!!

File: build/elasticsearch/bin/docker-entrypoint.sh [hunk 4]
```diff
@@ -59,4 +93,11 @@ if bin/elasticsearch-plugin list -s | grep -q x-pack; then
     fi
 fi
 
-exec bin/elasticsearch "${es_opts[@]}"
+if [[ "$(id -u)" == "0" ]]; then
+    # If requested and running as root, mutate the ownership of bind-mounts
+    if [[ -n "$MUTATE_MOUNTS" ]]; then
```
Comment (jarpy) [inline build/elasticsearch/bin/docker-entrypoint.sh]: `$TAKE_FILE_OWNERSHIP`? It's not obvious what kind of "mutation" is going to happen.

File: build/elasticsearch/bin/docker-entrypoint.sh (hunk 1 above)
Comment (dliappis) [reply build/elasticsearch/bin/docker-entrypoint.sh]: > From a different angle, did you consider changing the UID of the existing elasticsearch user instead? Is that a bad idea?

That's a good suggestion. One of my concerns was for people who want to inspect a running container and a new user (with something like `--user <random> -ti <container_id> /bin/bash`) .

But this is very much a corner case I think and testing it seems to work fine. Therefore I suggest an even bolder move: remove this entire section. I introduced it as suggested in the [Openshift docs](https://docs.openshift.org/latest/creating_images/guidelines.html) for fear of applications that want to lookup the associated user from `/etc/passwd`. I tried building things again now without it and every combination I tried doesn't give problems, including entering the container with an interactive shell. I also tried some of the binaries under `bin/` and `bin/x-pack` and didn't see issues.

Adding one test case that attempts to run elasticsearch with `--user=<random` will cover us somewhat, but naturally we can't write test cases to execute every binary in this scenario.

I vote for removing it altogether, WDYT?

File: build/elasticsearch/bin/docker-entrypoint.sh (hunk 2 above)
Comment (dliappis) [reply build/elasticsearch/bin/docker-entrypoint.sh]: Addressed in 9e3d43d20cbbeafea11a9fb4444cf27014255b90

File: build/elasticsearch/bin/docker-entrypoint.sh (hunk 4 above)
Comment (dliappis) [reply build/elasticsearch/bin/docker-entrypoint.sh]: Addressed in 74ca02ab848160ea57904307d1ed7594fbc085d6

File: build/elasticsearch/bin/docker-entrypoint.sh (hunk 1 above)
Comment (jarpy) [reply build/elasticsearch/bin/docker-entrypoint.sh]: 👍 to that. Less code is less complexity, and I'm always in favour of that.

Also, I think having an anonymous user that's identified only by UID makes it very clear that is a local, ephemeral user.

**Pull request merged**